docs/rule/no-mut-helper.md

```diff
@@ -0,0 +1,244 @@
+```
+
+```hbs
+<MyComponent @closeDropdown={{action this.setIsDropdownOpen false}}/>
+```
+
+```hbs
+<MyComponent onclick={{fn this.setIsDropdownOpen false}}/>
+```
+
+```hbs
+<Input onchange={{action this.setProfileDescription}}/>
+```
+
+Coupled with a corresponding JS action to set:
+
+```js
+@action
+setProfileDescription({ target: { value } }) {
+  set(this, 'profile.description', value);
+}
+```
+
+```hbs
+{{my-component click=(action (set this "setIsDropdownOpen" false)}}
+```
+
+```hbs
+{{my-component value=this.profile.description}}
+```
+
+## Migration
+
+1. When used as a pure setter only, `mut` could be replaced by a JS action or [ember-set-helper](https://github.com/pzuraq/ember-set-helper):
+
+Before:
+
+```hbs
+<MyComponent
+  @closeDropdown={{action (mut this.setIsDropdownOpen) false}}
+/>
+```
+
+After (Option 1 HBS):
+
+```hbs
+<MyComponent
+  @closeDropdown={{action this.setIsDropdownOpen false}}
+/>
+```
+
+After (Option 1 JS):
+
+```js
+@action
+setIsDropdownOpen(isDropdownOpen) {
+  set(this, 'isDropdownOpen', isDropdownOpen);
+}
+```
+
+After (Option 2):
+
+```hbs
+<MyComponent
+  @closeDropdown={{set this "isDropdownOpen" false}}
+/>
+```
+
+\
+2. When used as a pure getter only, `mut` could be removed:
+
+Before:
+
+```hbs
+<MyComponent
+  @isDropdownOpen={{mut this.isDropdownOpen}}
+/>
+```
+
+After:
+
+```hbs
+<MyComponent
+  @isDropdownOpen={{this.isDropdownOpen}}
+/>
+```
+
+\
+3. When `mut` is used as a getter and setter, `mut` could be replaced with a different namespace for the property and a dedicated action function to set the property: (Note: another other option could be to pull in the pick helper from [ember-composable-helpers](https://github.com/DockYard/ember-composable-helpers) and use it like [this](https://github.com/pzuraq/ember-set-helper#picking-values-with-ember-composable-helpers).) (Note: Another option could be to use [ember-box](https://github.com/pzuraq/ember-box)).
+
+Before:
+
+```hbs
+{{#let (mut this.foo) as |foo|}}
+  {{foo}}
+  <input onchange={{action foo value=”target.value”}} />
+{{/let}}
+```
+
+After HBS:
+
+```hbs
+{{this.foo}}
+<input {{on “change” this.updateFoo}} />
+```
+
+After JS:
+
+```js
+@tracked
+foo;
+
+@action
+updateFoo(evt) {
+  this.foo = evt.target.value;
+  // or set(this, ‘foo’, evt.target.value); for legacy Ember code
+}
+```
+
+\
+4. When `mut` is being passed into a built-in classic component that uses 2 way data binding, `mut` could be removed:
+
+Before:
+
+```hbs
+<Input
+  @value={{mut this.profile.description}}
+/>
+```
+
+After:
+
+```hbs
+<Input
+  @value={{this.profile.description}}
+/>
+```
+
+## Configuration
+
+* object -- containing the following properties:
+  * string -- `setterAlternative` -- Optional: String name of a helper that could replace mut as a setter. If configured, the lint violation error message will include this as a possible alternative for resolving the lint violation.
```

Comment:
I know we don't normally do this (though I think we should start adding it), but would you mind adding a small example config just below?

I'm thinking something like:

---

## Configuration

* object -- containing the following properties:
  * string -- `setterAlternative` -- Optional: String name of a helper that could replace mut as a setter. If configured, the lint violation error message will include this as a possible alternative for resolving the lint violation.

Example:

```js
// .template-lintrc.js

module.exports = {
  rules: {
    'no-mut-helper': {
      setterAlternative: 'what goes here?!?!?!?!',
    }
  }
};
```

Reply:
Also, doesn't this need to accept `boolean` too?

Reply:
thanks for the suggestion, Rob.

Agreed - i added a config example.

The config only expects an optional `string` (no `boolean`), which would be the name of helper that can be added to give engineers an extra hint on how to resolve the error, if their application defines or depends on an hbs helper alternative (like ember-set-helper$set) that could be used instead to resolve the violation without having to write a new JS action. If the config or the config.setterAlternative is undefined, then the default error message will be shown, which suggests engineers replace `mut` with a JS action. If the config accidentally passes a non-string value for config.setterAlternative, this is considered a mis-configuration and the rule throws an error to help the engineer fix the config.